src/Components/Popover/PopoverDialog.tsx

```diff
@@ -0,0 +1,429 @@
+}
+
+interface Props extends StandardProps {
+  /** Bounding rectangle of trigger element */
+  triggerRect: ClientRect | DOMRect
+  /** Pixel distance between trigger and popover dialog */
+  distanceFromTrigger: number
+  /** Where to position the popover relative to the trigger (assuming it fits there) */
+  position: PopoverPosition
+  /** Determins whether dialog is visible */
+  visible: boolean
+  /** Popover dialog color */
+  color: ComponentColor
+  /** Means of applying color to popover */
+  type: PopoverType
+  /** Popover dialog contents */
+  contents: JSX.Element
+  /** Handles clicks detected outside the popover dialog element */
+  onClickOutside: (e: MouseEvent) => void
+  /** Handles mouseleave events */
+  onMouseLeave: (e: MouseEvent) => void
+  /** Size of caret (triangle) that points at the trigger */
+  caretSize: number
+  /** Adds reasonable styles to popover dialog contents so you do not have to */
+  enableDefaultStyles: boolean
+}
+
+export class PopoverDialog extends Component<Props> {
+  public static readonly displayName = 'PopoverDialog'
+
+  private dialogRef = createRef<HTMLDivElement>()
+  private caretRef = createRef<HTMLDivElement>()
+
+  componentDidUpdate() {
+    const {dialogStyles, caretStyles} = this.popoverStyles
+
+    const dialogStyleString = convertCSSPropertiesToString(dialogStyles)
+    const caretStyleString = convertCSSPropertiesToString(caretStyles)
+
+    if (this.dialogRef.current) {
+      this.dialogRef.current.setAttribute('style', dialogStyleString)
+    }
+
+    if (this.caretRef.current) {
+      this.caretRef.current.setAttribute('style', caretStyleString)
+    }
+  }
+
+  render() {
+    const {
+      testID,
+      id,
+      style,
+      visible,
+      contents,
+      onClickOutside,
+      onMouseLeave,
+    } = this.props
+
+    if (!visible) {
+      return null
+    }
+
+    return (
+      <ClickOutside onClickOutside={onClickOutside}>
+        <div
+          onMouseLeave={onMouseLeave}
+          className={this.className}
+          ref={this.dialogRef}
+          data-testid={testID}
+          id={id}
+        >
+          <div
+            style={style}
+            className={this.contentsClassName}
+            data-testid={`${testID}--contents`}
+          >
+            {contents}
+          </div>
+          <div className="cf-popover--caret" ref={this.caretRef} />
+        </div>
+      </ClickOutside>
+    )
+  }
+
+  private get className(): string {
+    const {color, type, className} = this.props
+
+    return classnames('cf-popover', {
+      [`${className}`]: className,
+      [`cf-popover__${color}`]: color,
+      [`cf-popover__${type}`]: type,
+    })
+  }
+
+  private get contentsClassName(): string {
+    const {enableDefaultStyles} = this.props
+
+    return classnames('cf-popover--contents', {
+      'cf-popover--contents__default-styles': enableDefaultStyles,
+    })
+  }
+
+  private calculateDialogPosition = (): PopoverPosition => {
```

Comment:
I think that this method (`calculateDialogPosition`) along with `isDialogFlush` and `popoverStyles` should be utility functions rather than methods on this class.

As functions their dependencies are explicit (arguments), but as methods they have access to everything on `this`. I think when dependencies are made explicit, it opens up opportunities to refactor functions. And these are really gnarly, so any refactor here would be a win imo.